processing/src/main/java/io/druid/query/QueryContexts.java

```diff
@@ -132,6 +132,29 @@
     }
   }
 
+  public static <T> Query<T> verifyMaxQueryTimeout(Query<T> query, long maxQueryTimeout)
+  {
+    Object obj = query.getContextValue(TIMEOUT_KEY);
```

Comment:
This should be `getTimeout(query)` because otherwise it will not accept Strings (which `getTimeout` goes through pains to accept).

Reply:
Also, please modify `getDefaultTimeout(Query)` to be the min of `parseLong(query, DEFAULT_TIMEOUT_KEY, DEFAULT_TIMEOUT_MILLIS)` and the new `maxQueryTimeout` setting.

Reply:
Thanks @gianm, Changing verifyMaxQueryTimeout() to use `getTimeout(query)`. 

Passing `maxQueryTimeout ` in `getDefaultTimeout()` might be tricky b/c  I need to modify `getTimeout()` to take `maxQueryTimeout` as parameter b/c of.
```
  public static <T> long getTimeout(Query<T> query)
  {
    return getTimeout(query, getDefaultTimeout(query));
  }
```
 Alternative approach to handle passing ServerConfig values in QueryContext is to inject ServerConfig in the QueryContext class, for that I need to move QueryContext out of processing module. What do you think?

Reply:
Hmm, QueryContexts should stay in the processing module, I think.

I think it makes sense to pass the max timeout into `getTimeout`, however since it's annotated with `@PublicApi` you shouldn't change the signature. Instead you can `@Deprecate` it, add a new method with the proper signature, add a comment to the old method about its replacement, and move any internal callers to the new method.

Reply:
@gianm Thanks for the response. 
On
`move any internal callers to the new method.`
Some of the callers of `getTimeout()`(GroupByMergingQueryRunnerV2, DirectDruidClient, GroupByStrategyV2 and couple more) are in processing module and don't have access to ServerConfig. Should I move this configuration to some other config class which is available in `processing` and `server` module ?

Reply:
@akashdw Would it work to have a decorating QueryRunner that is added by the servers and which overrides the `timeout` context parameter as appropriate?

Reply:
👍 thanks @gianm , My initial plan was to have QueryRunner where we can do checks for `timeout`, `maxScatterGatherBytes` initialization, `maxTimeout` checks or any checks/validation in the future. Creating a separate PR for the same, once merged then I will add `maxTimeout` validation in the same QueryRunner.

Reply:
@gianm added `SetAndVerifyContextQueryRunner`. This is being added by Broker, Historical and Peon servers.